File: mocks/delivery-mock.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');

const app = express();
app.use(express.json());

const deliveries = new Map();

app.post('/api/v1/deliveries', (req, res) => {
  const { package: pkg, destination, window } = req.body;
  
  const deliveryId = `del_${uuidv4().substring(0, 8)}`;
  const droneId = `drone-${Math.floor(Math.random() * 20) + 1}`;
  
  const shouldFail = Math.random() < 0.08;
  
  if (shouldFail) {
    return res.status(400).json({
      error: 'WEATHER_RESTRICTION',
      message: 'High winds preventing drone flights'
    });
  }
  
  const eta = new Date();
  eta.setMinutes(eta.getMinutes() + 20);
  
  deliveries.set(deliveryId, {
    deliveryId,
    droneId,
    package: pkg,
    destination,
    status: 'IN_TRANSIT',
    eta: eta.toISOString(),
    location: {
      lat: 40.7128,
      lon: -74.0060
    }
  });
  
  setTimeout(() => {
    const delivery = deliveries.get(deliveryId);
    if (delivery) {
      delivery.status = 'DELIVERED';
      delivery.deliveredAt = new Date().toISOString();
    }
  }, 10000);
  
  res.status(201).json({
    deliveryId,
    droneId,
    eta: eta.toISOString()
  });
});

app.get('/api/v1/deliveries/:deliveryId', (req, res) => {
  const delivery = deliveries.get(req.params.deliveryId);
  
  if (!delivery) {
    return res.status(404).json({
      error: 'DELIVERY_NOT_FOUND',
      message: 'Delivery not found'
    });
  }
  
  res.json(delivery);
});

const PORT = process.env.PORT || 8083;
app.listen(PORT, () => {
  console.log(`Delivery mock service running on port ${PORT}`);
});

File: mocks/delivery_mock.py
"""
Mock Delivery Service
Simulates drone delivery with variable timing based on address distance.
"""
import random
import uuid
from datetime import datetime, timedelta
from fastapi import FastAPI
import uvicorn

app = FastAPI(title="Delivery Mock Service")

deliveries = {}

DRONE_IDS = [f"drone-{i}" for i in range(1, 16)]


@app.post("/api/v1/deliveries")
async def schedule_delivery(request: dict):
    if random.random() < 0.08:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail="Address not in delivery zone")

    delivery_id = f"del_{uuid.uuid4().hex[:8]}"
    drone_id = random.choice(DRONE_IDS)

    eta_minutes = random.randint(15, 45)
    eta = datetime.utcnow() + timedelta(minutes=eta_minutes)

    deliveries[delivery_id] = {
        "deliveryId": delivery_id,
        "status": "SCHEDULED",
        "droneId": drone_id,
        "destination": request.get("destination", {}),
        "package": request.get("package", {}),
        "eta": eta.isoformat() + "Z",
        "createdAt": datetime.utcnow().isoformat() + "Z",
        "location": None,
    }

    return {
        "deliveryId": delivery_id,
        "droneId": drone_id,
        "eta": eta.isoformat() + "Z",
    }


@app.get("/api/v1/deliveries/{delivery_id}")
async def get_delivery(delivery_id: str):
    if delivery_id not in deliveries:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail=f"Delivery {delivery_id} not found")

    delivery = deliveries[delivery_id]

    created = datetime.fromisoformat(delivery["createdAt"].replace("Z", ""))
    elapsed = (datetime.utcnow() - created).total_seconds()

    if elapsed > 5 and delivery["status"] == "SCHEDULED":
        delivery["status"] = "IN_TRANSIT"
        delivery["location"] = {"lat": 39.7817, "lng": -89.6501}

    if elapsed > 15 and delivery["status"] == "IN_TRANSIT":
        delivery["status"] = "DELIVERED"

    return delivery


@app.get("/health")
async def health():
    return {"status": "ok", "service": "delivery-mock"}


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8083)
